File: leetcode3/남효정/315. Count of Smaller Numbers After Self.py
# 풀이 실패
# 병합 정렬할 때 오른쪽에서 넘어간 개수 카운트하면 구할 수 있음
class Solution:
    def countSmaller(self, nums: List[int]) -> List[int]:
        n = len(nums)
        ans = [0] * n
        arr = [(nums[i], i) for i in range(n)]

        def merge_sort(enum):
            half = len(enum) // 2
            if half:
                left, right = merge_sort(enum[:half]), merge_sort(enum[half:])
                m, n = len(left), len(right)
                i = j = 0

                # 투 포인터 사용해서 병합 진행한다
                while i < m or j < n:
                    # 오른쪽 다 썼거나 왼쪽 값이 더 작거나 같은 경우
                    if j == n or (i < m and left[i][0] <= right[j][0]):
                        ans[left[i][1]] += j
                        enum[i + j] = left[i]
                        i += 1
                    else:
                        enum[i + j] = right[j]
                        j += 1

            return enum
            
        merge_sort(arr)
        return ans

        # 시간 복잡도 초과
        # ans = []
        # for i in range(len(nums)):
        #     if i == len(nums) - 1:
        #         ans.append(0)
        #         break
        #     check = 0
        #     for j in range(i + 1, len(nums)):
        #         if nums[i] > nums[j]:
        #             check += 1
        #     ans.append(check)
        
        # return ans

File: leetcode3/염혜정/315. Count of Smaller Numbers After Self.java
// o(nlogn)

class Solution {
    public List<Integer> countSmaller(int[] nums) {
        int n = nums.length;
        int[] counts = new int[n];
        int[] indices = new int[n];
        for (int i = 0; i < n; i++) indices[i] = i;
        mergeSort(nums, indices, counts, 0, n - 1);
        List<Integer> result = new ArrayList<>();
        for (int c : counts) result.add(c);
        return result;
    }

    private void mergeSort(int[] nums, int[] indices, int[] counts, int left, int right) {
        if (left >= right) return;
        int mid = (left + right) >>> 1;
        mergeSort(nums, indices, counts, left, mid);
        mergeSort(nums, indices, counts, mid + 1, right);
        merge(nums, indices, counts, left, mid, right);
    }

    private void merge(int[] nums, int[] indices, int[] counts, int left, int mid, int right) {
        int[] temp = new int[right - left + 1];
        int i = left, j = mid + 1, k = 0, rightCount = 0;
        while (i <= mid && j <= right) {
            if (nums[indices[j]] < nums[indices[i]]) {
                rightCount++;
                temp[k++] = indices[j++];
            } else {
                counts[indices[i]] += rightCount;
                temp[k++] = indices[i++];
            }
        }
        while (i <= mid) {
            counts[indices[i]] += rightCount;
            temp[k++] = indices[i++];
        }
        while (j <= right) temp[k++] = indices[j++];
        for (int p = 0; p < temp.length; p++) indices[left + p] = temp[p];
    }
}
